INSVCTwoFluidStaggeredHierarchyIntegrator.cpp

```diff
@@ -313,142 +335,136 @@ INSVCTwoFluidStaggeredHierarchyIntegrator::initializeLevelDataSpecialized(Pointe
                                                                           Pointer<BasePatchLevel<NDIM>> old_level,
                                                                           bool allocate_data)
 {
-    // Here is where we set initial data for our state variables.
-    // These are solvent velocity, network velocity, and pressure.
-    // NOTE: Pressure isn't really a "state" variable, but we let the data persist for initial guesses.
-    if (initial_time)
-    {
-        VariableDatabase<NDIM>* var_db = VariableDatabase<NDIM>::getDatabase();
-        const int un_sc_idx = var_db->mapVariableAndContextToIndex(d_un_sc_var, d_ctx);
-        const int us_sc_idx = var_db->mapVariableAndContextToIndex(d_us_sc_var, d_ctx);
-        const int p_cc_idx = var_db->mapVariableAndContextToIndex(d_p_cc_var, d_ctx);
-        Pointer<PatchLevel<NDIM>> level = hierarchy->getPatchLevel(level_number);
-        if (!level->checkAllocated(un_sc_idx)) level->allocatePatchData(un_sc_idx, init_data_time);
-        if (!level->checkAllocated(us_sc_idx)) level->allocatePatchData(us_sc_idx, init_data_time);
-        if (!level->checkAllocated(p_cc_idx)) level->allocatePatchData(p_cc_idx, init_data_time);
-
-        // Now let's set the initial data.
-        d_un_init_fcn->setDataOnPatchLevel(un_sc_idx, d_un_sc_var, level, init_data_time, true);
-        d_us_init_fcn->setDataOnPatchLevel(us_sc_idx, d_us_sc_var, level, init_data_time, true);
-        d_p_init_fcn->setDataOnPatchLevel(p_cc_idx, d_p_cc_var, level, init_data_time, true);
-
-        // Drawing data
-        if (d_visit_writer)
-        {
-            const int un_draw_idx = var_db->mapVariableAndContextToIndex(d_un_draw_var, d_ctx);
-            const int us_draw_idx = var_db->mapVariableAndContextToIndex(d_us_draw_var, d_ctx);
-            if (!level->checkAllocated(un_draw_idx)) level->allocatePatchData(un_draw_idx, init_data_time);
-            if (!level->checkAllocated(us_draw_idx)) level->allocatePatchData(us_draw_idx, init_data_time);
-        }
-    }
+    // Do any kind of Hierarchy specific initialization on a given patch level.
+    // Note: All initialization and regridding of state variables is managed by HierarchyIntegrator.
+    // Example uses of this include mantaining a vorticity value to see where things should be refined or checking norms
+    // of quantities before and after regridding. intentionally blank.
+}
+
+void
+INSVCTwoFluidStaggeredHierarchyIntegrator::applyGradientDetectorSpecialized(
+    const Pointer<BasePatchHierarchy<NDIM>> hierarchy,
+    const int level_num,
+    const double error_data_time,
+    const int tag_idx,
+    const bool initial_time,
+    const bool uses_richardson_extrapolation_too)
+{
+    // Fill in the tag_idx with 1 if the cell index on a given level should be refined.
+    // tag_idx corresponds to patch data CellData<NDIM, int>.
+    // TODO: Implement reasonable refinement criteria.
+}
```

Comment:
@bindi-nagda, this is the function that decides where to refine cells. It provides the patch data index `tag_idx` which corresponds to `CellData<NDIM, int>`. It should be filled in with 1 on cells that should be refined.

You should loop over the patches on the provided level number, and tag cells based on the gradient of theta.

Reply:
I've pushed some code for this